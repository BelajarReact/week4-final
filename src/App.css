body {
  font-family: Open Sans,sans-serif;
  font-size: 14px;
  font-weight: 200;
  color: #4a4a4a;
  background-color: #f7f7f7;
  -webkit-font-smoothing: antialiased;
}
.App{
  width: 100%;
  margin: 0 auto;
  min-height: 500px;
}
.navbar{
  background: rgb(0, 174, 239);
  position: fixed;
  width: 100%;
  top: 0px;
  left: 0px;
  right: 0px;
  z-index: 11;
}
.container{
  max-width: 480px;
}
.navbar a{
  vertical-align: middle;
}
.navbar input{
  width: auto;

  margin-right: 10px;
  flex: 0 0 auto;
  flex-grow: 1;
  display: flex;
  -webkit-box-pack: justify;
  justify-content: space-between;
  -webkit-box-align: center;
  align-items: center;
  color: rgb(255, 255, 255);
  font-size: 12px;
  height: 36px;
  background: rgb(27, 186, 245);
  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  padding: 0px 15px;
  border-radius: 20px;
}
.navbar input::placeholder{
  color: rgba(255,255,255,0.6);
}
.headerButton{

  align-items: center;
  display: flex;
  -webkit-box-pack: justify;
  justify-content: space-between;
  -webkit-box-align: center;
  align-items: center;
  color: rgb(255, 255, 255);
  font-size: 12px;
  height: 36px;
  width: auto;
  flex: 0 0 auto;
  flex-grow: 1;

  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  padding: 0px 15px;
  border-radius: 20px;
}

.AppContainer{
  background-color: white;
  position: relative;

  width: 100%;
  box-sizing: border-box;
  min-height: calc(100vh - 60px);
  margin: 0px auto;
  margin-top: 60px;
  padding: 20px 16px 330px;
}
.btn,.form-control{
  border-radius: 100px;
}

.heading{
  font-weight: bold;
  font-size: 20px;
  margin-bottom: 20px;
  text-align: center;
}

/*Loader*/
.search-loading {
  position: absolute;
  left: 0;
  right: 0;
  margin: auto;
}
/*Show and hide*/
.show {
  display:block;
}

.hide {
  display: none;
}

/**
REsults images
 */
.result-images.row{
  margin: 30px 0;
  margin-right: -10px;
  margin-left: -10px;
}
.result-images.row > .kolom{
  padding-right: 10px;
  padding-left: 10px;
}


.result-images .result-item {
  display: block;
  width: 100%;
  position: relative;

  height: 200px;
  overflow: hidden;
  background-color: #ddd;
  background-size: auto;
  background-repeat: no-repeat;
  background-position: center center ;
  background-attachment: inherit;
  border-radius: 4px;
  margin-bottom: 15px;
  cursor: pointer;
}
.result-images .result-item:after{
  content: "";
  position: absolute;
  width: 100%;
  height: 100%;
  top:0;
  left: 0;
  background-color: rgba(0,0,0,0.3);
  z-index: 1;
}
.image-author{
  position: absolute;
  display: inline-flex;
  padding: 3px 6px;
  font-size: small;
  background-color: red;
  border-radius: 6px;
  color:#fff;
  right: 10px;
  top:6px;
  z-index: 2;
  align-items: center;
}
.image-author i{
  margin-right: 6px;
}
.nav-link-container{
  display: flex;
}
.nav-link-container.hide{
  display: none;
}

.btn.disabled{
  cursor: no-drop;
}

.footer{
 position: absolute;
  bottom:0;
  text-align: center;
  line-height: 1.5;
  border-top: 1px solid #ddd;
  padding: 20px 0 15px;

}
.footer p{
  margin-bottom: 5px;
}
.footer .myLogo{
  background-color: #111;
  display: inline-block;
  padding: 10px;
  border-radius: 50px;
}
.footer .myLogo:hover{
  background-color: #4a4a4a;
  transform: scale(1.2);
}
.myLogo.white svg{
  fill: #fff;
}
.social-media a{
   display: inline-block;
  padding: 10px;
}
.stopwatch .time{
  text-align: center;
}
.stopwatch .time span{
  display: block;
  font-weight: bold;
}

.time span:first-child{
  font-size: 3rem;

}
.time span:last-child{
  text-transform: uppercase;
  font-size: 13px;
}

.btn-timer .btn{
  background-color: #4a4a4a;
  color:#fff;
  border-radius: 100%;
  line-height: 1;
  padding: 10px 12px;
}

.multisecond.fill{
  color: red;
}
.hours.fill{
  color: #1d4da7;
}
.minutes.fill{
  color: #06106d;
}

.seconds.fill{
  color: #3f4fde;
}