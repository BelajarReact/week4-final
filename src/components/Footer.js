import React, {Component} from "react";
import {Link} from "react-router-dom"
import Logo from "../logo.svg"

class SocialMedia extends Component{


    render() {
        const {socials} = this.props;

        return(

               <div className="social-media d-flex align-items-center justify-content-center">
                   {socials.map((item,index) => <a key={index} href={item.link} title={`Follow me at ${item.type}`} target="_blank"><i className={`fa ${item.icon}`}></i></a>)}
               </div>

        )
    }
}

const Footer = () => {

    const socialMedias = [
        {
            type: "facebook", icon: "fa-facebook-official", link: "https://fb.com/hadie.danker"
        },
        {
            type: "github", icon: "fa-github", link: "https://github.com/BelajarReact"
        },
        {
            type: "twitter", icon: "fa-twitter", link: "https://twitter.com/hadie87"
        },
        {
            type: "instagram", icon: "fa-instagram", link: "https://instagram.com/hadie.danker"
        },
        {
            type: "youtube", icon: "fa-youtube-play", link: "https://www.youtube.com/user/dankerizer"
        }
    ]

    const Pages  = [
        {
            link : "/",
            title :'Home page'
        },
        {
            link : "/about",
            title :'About'
        },
        {
            link : "/privacy",
            title :'Privacy'
        },
    ]
    return (
        <footer className="footer">
            <a href="https://fb.me/hadie.danker" className="myLogo white" target="_blank"><img src={Logo} height="30px"
                                                                                               width="30px" alt=""/></a>
            <p style={{marginTop: "20px"}}><i className="fa fa-user-circle-o"></i> Muhammad Syamsul Hadi</p>
            <p>Ini adalah final project dari kelas React FrontEnd by SanberCode.com</p>

            <p>Follo our social Media</p>

            <SocialMedia socials={socialMedias}/>
            <ul className="nav justify-content-center">
                {Pages.map((item,key) => <li className="nav-item" key={key}><Link className="nav-link" to={item.link} >{item.title}</Link></li>)}
            </ul>
        </footer>
    )
}

export default Footer;