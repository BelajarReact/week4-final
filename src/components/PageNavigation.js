import React from 'react';

export default ( props ) => {
    const {
        loading,
        showPrevLink,
        showNextLink,
        handlePrevClick,
        handleNextClick,
        hasQuery
    } = props;
    return (
        <div className={`nav-link-container btn-group btn-group-sm  ${ (showPrevLink || showNextLink) && !hasQuery ? 'show':'hide'}`}>
            <a
                href="#"
                className={
                    ` btn btn-secondary 
					${ showPrevLink ? '' : 'disabled'}
					${ loading ? 'greyed-out' : ''
                    }`
                }
                onClick={ handlePrevClick }
            >
                Prev
            </a>
            <a
                href="#"
                className={
                    ` btn btn-secondary
					${ showNextLink ? '' : 'disabled'}
					${ loading ? 'greyed-out' : '' }
					`}
                onClick={ handleNextClick }
            >
                Next
            </a>
        </div>
    )
}