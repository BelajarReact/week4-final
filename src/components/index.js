export {default as Waktu} from "./Waktu"
export {default as Search} from "./Search"
