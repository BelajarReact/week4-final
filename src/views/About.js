import React,{Component} from "react";

class About extends Component{

    render() {
        return(
            <div>
                <h1 style={{fontWeight:"bold",textAlign:"center",marginBottom:"20px"}}>Tentang Saya</h1>
            <p>Ayah dari 3 jagoan, Arsy, Arsyad, Abdurrahman -> Ibnu Muhammad bin Suyuthi</p>
                <p>Lahir di desa kecil bernama Wonosari, besar di banyak tempat karena Ayahnya petani</p>
                <p>Sekarng dirumahAja, sebagai WordPress Developer</p>

                Lorem ipsum dolor sit amet, consectetur adipisicing elit. Accusamus assumenda, at commodi, delectus dolore dolorum esse, eum facilis incidunt ipsam itaque obcaecati officiis omnis pariatur quam quas quia rem soluta.
            </div>
        )
    }
}

export default About;