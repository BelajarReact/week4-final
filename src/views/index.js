export { default as Home } from "./Home";
export { default as LoginForm } from "./LoginForm";
export { default as StopWatch } from "./StopWatch";
export { default as About } from "./About";
export { default as Privacy } from "./Privacy";