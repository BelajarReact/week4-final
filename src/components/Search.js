import React from 'react';
// import '../Search.css';
import axios from 'axios';
import Loader from './loader.svg';
import Pixabay from './pixabay.svg';
import PageNavigation from './PageNavigation';

class Search extends React.Component {

    constructor( props ) {
        super( props );

        this.state = {
            query: '',
            results: {},
            loading: false,
            message: '',
            totalResults: 0,
            totalPages: 0,
            currentPageNo: 0,
        };

        this.cancel = '';
    }


    /**
     * Get the Total Pages count.
     *
     * @param total
     * @param denominator Count of results per page
     * @return {number}
     */
    getPageCount = ( total, denominator ) => {
        const divisible	= 0 === total % denominator;
        const valueToBeAdded = divisible ? 0 : 1;
        return Math.floor( total/denominator ) + valueToBeAdded;
    };

    /**
     * Fetch the search results and update the state with the result.
     * Also cancels the previous query before making the new one.
     *
     * @param {int} updatedPageNo Updated Page No.
     * @param {String} query Search Query.
     *
     */
    ambilDataPixabay = ( updatedPageNo = '', query ) => {
        const pageNumber = updatedPageNo ? `&page=${updatedPageNo}` : '';
        const searchUrl = `https://pixabay.com/api/?key=446579-0544d1523568f48bffb749e2d&q=${query}${pageNumber}`;

        if( this.cancel ) {
            this.cancel.cancel();
        }

        this.cancel = axios.CancelToken.source();

        axios.get( searchUrl, {
            cancelToken: this.cancel.token
        } )
            .then( res => {
                const total = res.data.total;
                const totalPagesCount = this.getPageCount( total, 20 );
                const resultNotFoundMsg = ! res.data.hits.length
                    ? 'Gambar tidak ditemukan, silahkan gunakan kata kunci lainya'
                    : '';
                this.setState( {
                    results: res.data.hits,
                    message: resultNotFoundMsg,
                    totalResults: total,
                    totalPages: totalPagesCount,
                    currentPageNo: updatedPageNo,
                    loading: false
                } )
            } )
            .catch( error => {
                if ( axios.isCancel(error) || error ) {
                    this.setState({
                        loading: false,
                        message: 'Failed to fetch the data. Please check network'
                    })
                }
            } )
    };

    handleInputPencarian = ( event ) => {
        const query = event.target.value;
        if ( ! query ) {
            this.setState( { query, results: {}, message: '', totalPages: 0, totalResults: 0 } );
        } else {
            this.setState( { query, loading: true, message: '' }, () => {
                this.ambilDataPixabay( 1, query );
            } );
        }
    };


    handlePageClick = ( type ) => {
        // event.preventDefault();
        const updatePageNo = 'prev' === type
            ? this.state.currentPageNo - 1
            : this.state.currentPageNo + 1;

        if( ! this.state.loading  ) {
            this.setState( { loading: true, message: '' }, () => {
                this.ambilDataPixabay( updatePageNo, this.state.query );
            } );
        }
    };

    renderSearchResults = () => {
        const { results } = this.state;

        if ( Object.keys( results ).length && results.length ) {
            return (
                <div className="row result-images row-no-gap">
                    { results.map( result => {
                        return (
                            <div className="col-6 kolom" key={ result.id }>
                            <a  href={ result.previewURL }  target="_blank" className="result-item" style={{backgroundImage:`url(${result.webformatURL})`,display:"block"}}>
                                <span className="image-username image-author"><i className="fa fa-user-circle-o"  aria-hidden="true"></i>  {result.user}</span>



                            </a>
                            </div>
                        )
                    } ) }

                </div>
            )
        }
    };

    componentDidMount() {

    }

    componentWillUnmount() {

        this.setState({
            results:[]
        })
    }

    render() {
        const { query, loading, message, currentPageNo, totalPages } = this.state;

        const showPrevLink = 1 < currentPageNo;
        const showNextLink = totalPages > currentPageNo;

        return (
            <div className="container">
                {/*	Heading*/}
                <h2 className="heading">Pencarian Gambar dari Pixabay</h2>
                {/* Search Input*/}
                <label className="search-label" htmlFor="search-input" style={{display:"block"}}>
                    <input
                        type="text"
                        name="query"
                        value={ query }
                        id="search-input"
                        aria-label="search-input"
                        className="form-control"
                        placeholder="Search..."
                        onChange={this.handleInputPencarian}
                    />

                </label>

                {/*	Error Message*/}
                {message && <p className="message">{ message }</p>}

                {/*	Loader*/}
                <img src={ Loader } className={`search-loading ${ loading ? 'show' : 'hide' }`} alt="loader"/>

                {/*Navigation*/}
                <PageNavigation
                    loading={loading}
                    showPrevLink={showPrevLink}
                    showNextLink={showNextLink}
                    handlePrevClick={ () => this.handlePageClick('prev')}
                    handleNextClick={ () => this.handlePageClick('next')}
                />

                {/*	Result*/}
                { this.renderSearchResults() }

                {/*Navigation*/}
                <PageNavigation
                    loading={loading}
                    showPrevLink={showPrevLink}
                    hasQuery={ query === '' }
                    showNextLink={showNextLink}
                    handlePrevClick={ () => this.handlePageClick('prev' )}
                    handleNextClick={ () => this.handlePageClick('next' )}
                />

                <div  className={query ===''?'hide':'show'} style={{textAlign:"center",margin:"20px auto"}}>
                    <img src={Pixabay} height="30"  alt="" />
                </div>

            </div>
        )
    }
}

export default Search