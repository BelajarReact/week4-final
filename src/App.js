import React from 'react';
import './App.css';
import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";
import { Home, LoginForm } from "./views";
import Footer from "./components/Footer";
const App =() => {


  return (
    <div className="App">
        <Router>
        <nav className="navbar navbar-expand-xs navbar-dark navbar-fixed">
            <div className="container">
                <Link className="navbar-brand text-white" to="/"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 98.124 98.124" fill="#fff" height="30px"><defs/><path d="M92.727 63.157c-1.18.547-1.137 3.172-2.958 5.723-2.335 3.269-5.874 3.521-8.086 1.787-9.121-7.152 9.048-38.204-3.79-43.623-13.135-5.544-22.26 11.128-29.646 23.812 2.374-13.763 9.587-29.608-2.922-34.795-10.969-4.548-25.861 14.91-34.874 40.42C10.6 44.244-.282 46.935.006 49.915c.187 1.92 3.188 10.692 2.729 28.473-.063 2.54 6.558 7.064 9.139 1.93 4.027-8.011 14.521-38.48 28.534-53.359-3.496 17.824-5.229 38.386-5.052 50.746.036 2.479 6.223 7.084 9.434 1.696 3.388-5.684 13.339-30.289 27.707-42.675-5.425 18.34-8.897 38.096 2.855 44.085 11.896 6.062 21.64-3.554 22.698-11.067.521-3.704-1.759-8.238-5.323-6.587z"/></svg>
                    </Link>
                    <div className="headerButton">
                        <input type="text" className="form-control form-control-sm" placeholder="Search here..."/>
                        <button className="navbar-toggler" type="button" data-toggle="collapse"
                                data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false"
                                aria-label="Toggle navigation">
                            <span className="navbar-toggler-icon"></span>
                        </button>
                    </div>

                <div className="collapse navbar-collapse" id="navbarSupportedContent">
                    <ul className="navbar-nav mr-auto">
                        <li className="nav-item active">
                            <Link className="nav-link" to="/">
                                Beranda
                            </Link>
                        </li>
                        <li className="nav-item">
                            <Link className="nav-link" to="/login">
                                Login Form
                            </Link>
                        </li>

                        <li className="nav-item">
                            <Link className="nav-link" to="/kalkulator">
                                Kalkulator
                            </Link>
                        </li>
                    </ul>

                </div>
            </div>
        </nav>


           <div className="container AppContainer">
               <Switch>
                   <Route exact path="/">
                     <Home/>
                   </Route>
                   <Route exact path="/login">
                       <LoginForm/>
                   </Route>
                   <Route exact path="/kalkulator">
                       ini adalah kalkulator
                   </Route>
               </Switch>

            <Footer/>
           </div>
       </Router>


    </div>
  );
}

export default App;
