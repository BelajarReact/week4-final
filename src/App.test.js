import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import App from './App';
import axios from 'axios';
import '@testing-library/jest-dom'
import {rest} from "msw"
import {setupServer} from 'msw/node'
import {Search} from "./components";
import { Home, LoginForm,StopWatch,About,Privacy } from "./views";

// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });
const searchUrl = `https://pixabay.com/api/?key=446579-0544d1523568f48bffb749e2d&q=car&page=1`;

const server = setupServer(
    rest.get(searchUrl, (req, res, ctx) => {
      return res(ctx.json({key: '446579-0544d1523568f48bffb749e2d'}))
    }),
)
beforeAll(() => server.listen())
afterEach(() => {
  server.resetHandlers()
  window.localStorage.removeItem('token')
})
afterAll(() => server.close())

const setupSearch = () => {
    const utils = render(<Search />)
    const input = utils.getByLabelText('search-input')
    return {
        input,
        ...utils,
    }
}

test("Test Search Form Pixabay",() =>{
    const { input } = setupSearch()
    fireEvent.change(input, { target: { value: 'car' } })

    expect(input.value).toBe('car')

    const component = renderer.create(
        <Search/>
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
})

test('stop watch event',() =>{
    const { getByTestId, rerender } = render(<StopWatch timerOn={true} />)
    expect(getByTestId('centiseconds').textContent).greatherThan(0)

})