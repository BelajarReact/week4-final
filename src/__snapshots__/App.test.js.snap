// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Search Form Pixabay 1`] = `
<div
  className="container"
>
  <h2
    className="heading"
  >
    Pencarian Gambar dari Pixabay
  </h2>
  <label
    className="search-label"
    htmlFor="search-input"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <input
      aria-label="search-input"
      className="form-control"
      id="search-input"
      name="query"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
  </label>
  
  <img
    alt="loader"
    className="search-loading hide"
    src="loader.svg"
  />
  <div
    className="nav-link-container btn-group btn-group-sm  hide"
  >
    <a
      className=" btn btn-secondary 
					disabled
					"
      href="#"
      onClick={[Function]}
    >
      Prev
    </a>
    <a
      className=" btn btn-secondary
					disabled
					
					"
      href="#"
      onClick={[Function]}
    >
      Next
    </a>
  </div>
  <div
    className="nav-link-container btn-group btn-group-sm  hide"
  >
    <a
      className=" btn btn-secondary 
					disabled
					"
      href="#"
      onClick={[Function]}
    >
      Prev
    </a>
    <a
      className=" btn btn-secondary
					disabled
					
					"
      href="#"
      onClick={[Function]}
    >
      Next
    </a>
  </div>
  <div
    className="hide"
    style={
      Object {
        "margin": "20px auto",
        "textAlign": "center",
      }
    }
  >
    <img
      alt=""
      height="30"
      src="pixabay.svg"
    />
  </div>
</div>
`;
